x/ibc-rate-limit/contracts/rate-limiter/src/contract.rs

```diff
@@ -0,0 +1,527 @@
+            if limit.flow.is_expired(now) {
+                limit.flow.expire(now, limit.quota.duration)
+            }
+            limit.flow.add_flow(direction.clone(), funds);
+
+            let balance = limit.flow.balance();
+            if balance > max {
+                return Err(ContractError::RateLimitExceded {
+                    channel: channel_id.to_string(),
+                    reset: limit.flow.period_end,
+                });
+            };
+            Ok(RateLimitResponse {
+                // TODO: nit, can we derive channel.Clone()?
+                rate_limit: RateLimit {
+                    quota: limit.quota.clone(),
+                    flow: limit.flow,
+                },
+                used: balance,
+                max,
+            })
+        })
+        .collect();
+    let results = results?;
+
+    RATE_LIMIT_TRACKERS.save(
+        deps.storage,
+        &channel_id,
+        &results.iter().map(|r| r.rate_limit.clone()).collect(),
+    )?;
+
+    let response = Response::new()
+        .add_attribute("method", "try_transfer")
+        .add_attribute("channel_id", channel_id);
+
+    // Adding the attributes from each quota to the response
+    // Code style Q: Should we move attribute setting to a function on response?
+    // Rust question: How does this work with one response being an error, I'm not sure how the flow works here
+    results.iter().fold(Ok(response), |acc, result| {
+        Ok(acc?
+            .add_attribute(
+                format!("{}_used", result.rate_limit.quota.name),
+                result.used.to_string(),
+            )
+            .add_attribute(
+                format!("{}_max", result.rate_limit.quota.name),
+                result.max.to_string(),
+            )
+            .add_attribute(
+                format!("{}_period_end", result.rate_limit.quota.name),
+                result.rate_limit.flow.period_end.to_string(),
+            ))
+    })
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::GetQuotas { channel_id } => get_quotas(deps, channel_id),
+    }
+}
+
+fn get_quotas(deps: Deps, channel_id: impl Into<String>) -> StdResult<Binary> {
+    to_binary(&RATE_LIMIT_TRACKERS.load(deps.storage, &channel_id.into())?)
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn migrate(_deps: DepsMut, _env: Env, _msg: MigrateMsg) -> Result<Response, ContractError> {
+    unimplemented!()
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use cosmwasm_std::testing::{mock_dependencies, mock_env, mock_info};
+    use cosmwasm_std::{from_binary, Addr, Attribute};
+
+    use crate::helpers::tests::verify_query_response;
+    use crate::msg::{Channel, QuotaMsg};
+    use crate::state::RESET_TIME_WEEKLY;
+
+    const IBC_ADDR: &str = "IBC_MODULE";
+    const GOV_ADDR: &str = "GOV_MODULE";
+
+    #[test]
```

Comment:
I wonder if we have something like the test suite in rust, maybe utilize helpers.rs, we should revisit to have test setup methods that we can reuse!

Reply:
yeah, I find cosmwasm tests to be quite repetitive. They're quick to write, but for larger suites it might be useful to abstract some of it.